[PL-49179] Update delegate expiration policy (#6441)
harness/developer-hub: docs/platform/delegates/install-delegates/delegate-upgrades-and-expiration.md

sidebar_position: 9
---

> The Harness Delegate for Kubernetes and Helm supports automatic upgrades. Harness recommends that you enable automatic upgrades for your Kubernetes delegates. With automatic upgrades, you don't need to worry about having the most up-to-date functionality. You always have the most recent version of the delegate.

Author: ```suggestion
The Harness Delegate for Kubernetes and Helm supports automatic upgrades. Harness recommends that you enable automatic upgrades for your Kubernetes delegates. With automatic upgrades, you always have the most recent version of the delegate.
```

## How automatic upgrade works in the Kubernetes manifest

> The Kubernetes manifest has a component called `upgrader`. The `upgrader` is a cron job that runs every hour by default. Every time it runs, it makes a call to Harness Manager to determine which delegate version is published for the account. The API returns a payload, such as `harness/delegate:yy.mm.verno`. If the delegate that was involved in this upgrade cron job does not have the same image as what the API returns, the `kubectl set image` command runs to perform a default rolling deployment of the delegate replicas with the newer image.

Reviewer: ```suggestion
The Kubernetes manifest has a component called `upgrader`. The `upgrader` is a cron job that runs every hour by default. Every time it runs, it sends a request to Harness Manager to determine which delegate version is published for the account. The API returns a payload, such as `harness/delegate:yy.mm.verno`. If the delegate that was involved in this upgrade cron job does not have the same image as what the API returns, the `kubectl set image` command runs to perform a default rolling deployment of the delegate replicas with the newer image.
```

    The next time the `upgrader` runs, it will receive the `artifactory-abc/harness/delegate:24.04.82804` image.

> ### Delegate expiration support policy

Author: ```suggestion
## Delegate expiration support policy
```

> Six months after a delegate image is released, the delegate reaches End of Support (EOS). Eight months after a delegate image is released, the delegate is End of Life (EOL). Delegates expire if not upgraded 6 months after the image is released.

Author: ```suggestion
Six months after a delegate image is released, the delegate reaches End of Support (EOS). Eight months after a delegate image is released, the delegate is End of Life (EOL). 

Delegates do not stop working after expiration. Because delegates are only backward-compatible, they might have issues if the backend has moved too far ahead. Harness recommends that you upgrade your delegates before they expire.
```

## Delegate expiration policy

> Delegates expire after 8 months from the release of the delegate image on Docker Hub, with 6 months of support and a 2 month upgrade period. For more information, go to [Expiration support policy](#expiration-support-policy).

Author: ```suggestion
Delegates expire after 8 months from the release of the delegate image on Docker Hub, with 6 months of support and a 2-month upgrade period. For more information, go to [Expiration support policy](#expiration-support-policy).
```

> Delegates do not stop working after expiration. Because delegates are only backward-compatible, they might have issues if the backend has moved too far ahead. Harness recommends that you upgrade your delegates before they expire. For more information about delegate releases, go to the [Delegate release notes](/release-notes/delegate).

Reviewer: ```suggestion
Delegates do not stop working after expiration. Because delegates are only backward-compatible, they might have issues if the back end has moved too far ahead. Harness recommends that you upgrade your delegates before they expire. For more information about delegate releases, go to the [Delegate release notes](/release-notes/delegate).
```

:::info important
If you do not have automatic upgrade enabled, Harness recommends upgrading the delegate at least once per quarter. You must update the delegate every 8 months.
:::

### Delegate expiration support policy

> Six months after a delegate image is released, the delegate reaches End of Support (EOS). Eight months after a delegate image is released, the delegate is End of Life (EOL). Delegates expire if not upgraded 8 months after the image is released.

Author: ```suggestion
Six months after a delegate image is released, the delegate reaches End of Support (EOS). Eight months after a delegate image is released, the delegate is End of Life (EOL). Delegates expire if not upgraded 8 months after the image is released. Delegates might not work after they reach EOS. When delegates reach EOS, Harness strongly recommends that you update.
```

> Six months after a delegate image is released, the delegate reaches End of Support (EOS). Eight months after a delegate image is released, the delegate is End of Life (EOL). Delegates expire if they are not upgraded 8 months after the image is released.

Author: ```suggestion
Six months after a delegate image is released, the delegate reaches End of Support (EOS). Eight months after a delegate image is released, the delegate is End of Life (EOL). Delegates expire if not upgraded 8 months after the image is released.
```

| Release | EOS | EOL |
| --- | --- | --- |
| 24.04.verno | 23.10.verno and below | 23.08.verno and below |
| 24.05.verno | 23.11.verno and below | 23.09.verno and below |

EOS means the following:

- Harness Support will no longer accept support requests for the delegate in both Harness FirstGen and Harness NextGen.

- Security fixes will still be addressed.

> - Product defects will not be addressed.

Author: ```suggestion
- Product defects will not be addressed.
- If delegates are past their EOS date, Harness does not support them. Expired delegates might not work as intended. For issues with expired delegates, Harness Support will request that you upgrade your delegate(s).
```

EOL means the following:

- In addition to the EOS clauses, security fixes will not be addressed.

> #### Example delegate expiration

Author: ```suggestion
#### Example delegate expiration

```

 For delegates with an immutable image type, the image tag is `yy.mm.verno`. A delegate version `24.05.84200` would reach EOS in November 2024 and EOL in January 2025.

:::note
This policy applies to delegates with the `yy.mm.verno` image tag. It does not apply to legacy delegates. For information on delegate types, go to [Delegate image types](/docs/platform/delegates/delegate-concepts/delegate-image-types).

:::

:::info

> For Harness Self-Managed Enterprise Edition, support is limited to the delegate version released with the most recent version. When you upgrade your Harness Self-Managed Enterprise Edition, the supported delegate version is included.

Author: ```suggestion
Harness Self-Managed Enterprise Edition support is limited to the delegate version released with the most recent version. When you upgrade Harness Self-Managed Enterprise Edition, the supported delegate version is included.
```
